enhancements/microshift/multus-cni-for-microshift.md

```diff
@@ -0,0 +1,542 @@
+will be prepared and will only include relevant artifacts such as entrypoint script/binary and
+Multus CNI binary (which is copied to host's `/opt/cni/bin`).
+
+To supply network plugins (CNIs) such as `bridge`, `ipvlan`, and `macvlan` a new image will be prepared
+so MicroShift uses the same binaries as OpenShift (alternative is using RHEL's networkplugins RPM - see alternatives).
+This image will also contain IPAM binaries such as `static`, `dynamic (DHCP)`, and `host-local`.
+
+Both of these images will be part of the OpenShift payload which MicroShift references during rebase procedure.
+
+### Risks and Mitigations
+
+There may be a race condition between Multus on MicroShift and other services on the host that also
+configure the host's networking
+Taking `bridge` CNI as an example: when the bridge interface does not exist, it will be created
+when a Pod requiring that interface is created.
+If user expects something else will create the interface, they will need to configure system to start
+MicroShift after other services.
+
+Multus using thin plugin architecture creates a kubeconfig when the DaemonSet starts and copies it
+out of Pod to `/etc/cni/net.d/multus.d/multus.kubeconfig` on the host filesystem so the Multus CNI
+binary can use it to get
+NetworkAttachmentDefinitions and Pod annotations. The file is owned and readable only by root, so
+the risk of someone gaining access to the cluster after logging into the host can be compared to
+getting access to `/var/lib/microshift/resources/kubeadmin/kubeconfig`.
+
+Currently MicroShift runs as a single-node cluster but in future releases there might be an effort
+to allow for multi-node clusters. For this reason, during implementation of this enhancement, no
+assumption that MicroShift will always run as a single-node should be made and potential multi-node
+should be kept in mind. We will not build the solution to support multi-node, but want to avoid
+making decisions that make it harder to do so in the future.
+
+Another potential risk was investigated but judged as not a problem: could a race condition between
+Multus being ready and application's Pods starting impact the application, by for example not having
+relevant parts of network set up.
+This should not be an issue because Pods that do not use `hostNetwork` will have networking setup
+after the CNI is ready. Both ovn-kubernetes and Multus use `hostNetwork` so they start before other
+Pods. If CRI-O is configured to use the Multus regardless of files in /etc/cni/net.d, CRI-O will wait
+for the Multus. We can also use the Multus' config option of `--readiness-indicator-file` to make
+sure the Multus waits for the ovn-kubernetes.
+
+Another existing risk is users creating NetworkAttachmentDefinitions that want to reuse MicroShift's
+networks such as `10.42.0.0/16`, `10.43.0.0/16`, `10.44.0.0/16`, or `169.254.169.1/29`.
+We should warn users they should not use them unless they reconfigure MicroShift's networking to use
```

Comment:
What happens if users don't add other networks and reuse the default settings?

Reply:
Hard for me to say exactly, someone from networking team would know better.
But if user reuses `10.42.0.0/16`, which is used by primary ovn-k CNI, then I would expect problems with routing - the Pod wouldn't be able to access ApiServer (if the second CNI would overwrite ovnk's routing), or some services if other CIDR is used.